theano/sandbox/rng_mrg.py

```diff
@@ -1029,96 +1048,177 @@ def multinomial_wo_replacement(self, size=None, n=1, pvals=None,
         else:
-            # if even, don't change, if odd, +1
-            n_samples = prod(size) + (prod(size) % 2)
-        flattened = self.uniform(size=(n_samples,), dtype=dtype,
-                                 nstreams=nstreams)
-
-        if constant:
-            U1 = flattened[:n_samples // 2]
-            U2 = flattened[n_samples // 2:]
-        else:
-            U1 = flattened[:prod(flattened.shape) // 2]
-            U2 = flattened[prod(flattened.shape) // 2:]
-
-        # normal_samples = zeros_like(flattened)
-        sqrt_ln_U1 = sqrt(-2.0 * log(U1))
-        # TypeError: 'TensorVariable' object does not support item assignment
-        # so this doesn't work...
-        # normal_samples[:n_samples/2] = sqrt_ln_U1 * cos(2.0*np.pi*U2)
-        # normal_samples[n_samples/2:] = sqrt_ln_U1 * sin(2.0*np.pi*U2)
-
-        # so trying this instead
-        first_half = sqrt_ln_U1 * cos(
-            np.array(2.0 * np.pi, dtype=dtype) * U2)
-        second_half = sqrt_ln_U1 * sin(
-            np.array(2.0 * np.pi, dtype=dtype) * U2)
-        normal_samples = join(0, first_half, second_half)
-
-        final_samples = None
-        if evened:
-            final_samples = normal_samples[:-1]
-        elif constant:
-            final_samples = normal_samples
-        else:
-            final_samples = normal_samples[:prod(size)]
+            norm_samples = tensor.join(0, z0, z1)
+
+        samples = norm_samples[:n_odd_samples]
+        samples = tensor.reshape(samples, newshape=size, ndim=ndim)
+        samples *= std
+        samples += avg
+
+        return samples
+
+    def truncated_normal(self, size, avg=0.0, std=1.0,
+                         ndim=None, dtype=None, **kwargs):
+        """
+        Sample a tensor of values from a symmetrically truncated normal distribution.
+
+        Parameters
+        ----------
+        size : int_vector_like
+            Array dimensions for the output tensor.
+        avg : float_like, optional
+            The mean value for the truncated normal to sample from (defaults to 0.0).
+        std : float_like, optional
+            The standard deviation for the truncated normal to sample from (defaults to 1.0).
+        ndim : int, optional
+            The number of dimensions for the output tensor (defaults to None).
+            This argument is necessary if the size argument is ambiguous on the number of dimensions.
+        dtype : str, optional
+            The data-type for the output tensor. If not specified,
+            the dtype is inferred from avg and std, but it is at least as precise as floatX.
+        kwargs
+            Other keyword arguments for random number generation (see uniform).
+
+        Returns
+        -------
+        samples : TensorVariable
+            A Theano tensor of samples randomly drawn from a truncated normal distribution.
+
+        See Also
+        --------
+        normal
+        """
+        std = std / tensor.constant(.87962566103423978)
```

Comment:
Is there a mathematical explanation/expression for that number (`.87962566103423978`) ? If so, could you use it instead of putting directly the number value ? As you do in line 1118 with `lower = tensor.constant(1. / np.e**2, dtype=dtype)`.

Reply:
It is the result of `scipy.stats.truncnorm.std(a=-2, b=2, loc=0., scale=1.)` (see comment in `test_normal_truncation`). I was not sure whether this is worth a scipy import or whether Theano already depends on scipy. Should I replace the constant by the above mentioned line?

Reply:
Theano has no hard dependency on scipy, so it's better not to import it here.
If you could just copy the comment here as well, I think that should be enough.

Reply:
I agree with @lamblin , just adding a comment to tell how it is calculated should be sufficient.

Reply:
PS: Ok, comment is in `test_normal_truncation`. Could you add it here too, in `truncated_normal()` function ?